src/Dapr.Client/DaprClientGrpc.cs

```diff
@@ -162,7 +162,96 @@ internal class DaprClientGrpc : DaprClient
             }
         }
 
+        /// <inheritdoc/>
+        public override Task<BulkPublishResponse> BulkPublishEventAsync<TValue>(
+            string pubsubName,
+            string topicName,
+            IReadOnlyList<TValue> events,
+            string contentType,
+            CancellationToken cancellationToken = default)
+        {
+            return BulkPublishEventAsync(pubsubName, topicName, events, contentType, null, cancellationToken);
+        }
+        
+        /// <inheritdoc/>
+        public override Task<BulkPublishResponse> BulkPublishEventAsync<TValue>(
+            string pubsubName,
+            string topicName,
+            IReadOnlyList<TValue> events,
+            string contentType,
+            Dictionary<string, string> metadata,
+            CancellationToken cancellationToken = default)
+        {
+            ArgumentVerifier.ThrowIfNullOrEmpty(pubsubName, nameof(pubsubName));
+            ArgumentVerifier.ThrowIfNullOrEmpty(topicName, nameof(topicName));
+            ArgumentVerifier.ThrowIfNull(events, nameof(events));
+            return MakeBulkPublishRequest(pubsubName, topicName, events, contentType, metadata, cancellationToken);
+        }
+        
+        private async Task<BulkPublishResponse> MakeBulkPublishRequest<TValue>(
+            string pubsubName,
+            string topicName,
+            IReadOnlyList<TValue> events,
+            string contentType,
+            Dictionary<string, string> metadata,
+            CancellationToken cancellationToken)
+        {
+            if (String.IsNullOrEmpty(contentType))
+            {
+                contentType = Constants.ContentTypeApplicationJson;
+            }
+            
+            var envelope = new Autogenerated.BulkPublishRequest()
+            {
+                PubsubName = pubsubName,
+                Topic = topicName,
+            };
+
+            for (int counter = 0; counter < events.Count; counter++)
+            {
+                var entry = new Autogenerated.BulkPublishRequestEntry()
+                {
+                    EntryId = counter.ToString(),
+                    Event = TypeConverters.ToJsonByteString(events[counter], this.jsonSerializerOptions),
```

Comment:
Actually, since we always convert the data to json, would we ever be a different content type? Do we need to support that?

Reply:
We only support json serialization for now in pubsub (normal publish events). We should probably introduce other serializers in utils and give users the flexibility to choose the content type of their type.